File: app/src/main/java/com/gproust/sprout/ui/common/Components.kt
@file:OptIn(ExperimentalMaterial3Api::class, ExperimentalLayoutApi::class)

package com.gproust.sprout.ui.common

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material.icons.filled.ExpandLess
import androidx.compose.material.icons.filled.ExpandMore
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Card
import androidx.compose.material3.CenterAlignedTopAppBar
import androidx.compose.material3.DatePicker
import androidx.compose.material3.DatePickerDefaults
import androidx.compose.material3.DatePickerDialog
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.FilterChip
import androidx.compose.material3.FloatingActionButton
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.SelectableDates
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TimePicker
import androidx.compose.material3.rememberDatePickerState
import androidx.compose.material3.rememberTimePickerState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.ui.window.Dialog
import com.gproust.sprout.R
import java.time.Instant
import java.time.LocalDate
import java.time.LocalTime
import java.time.ZoneId
import java.time.ZoneOffset

@Composable
fun SproutTopBar(title: String, onBack: (() -> Unit)? = null) {
    CenterAlignedTopAppBar(
        title = { Text(title) },
        navigationIcon = {
            if (onBack != null) {
                IconButton(onClick = onBack) {
                    Icon(
                        Icons.AutoMirrored.Filled.ArrowBack,
                        contentDescription = stringResource(R.string.cd_back),
                    )
                }
            }
        },
    )
}

@Composable
fun StatCard(label: String, value: String, icon: ImageVector, modifier: Modifier = Modifier) {
    Card(modifier = modifier) {
        Column(Modifier.padding(16.dp)) {
            Icon(icon, contentDescription = null, tint = MaterialTheme.colorScheme.primary)
            Text(
                value,
                style = MaterialTheme.typography.titleLarge,
                modifier = Modifier.padding(top = 8.dp),
            )
            Text(
                label,
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
    }
}

/**
 * One logged entry in a history list. [details] adds an expandable breakdown
 * under the row; [action] puts a one-tap shortcut next to it (e.g. "Mark as
 * used"), for the state change that would otherwise mean opening the editor.
 */
@Composable
fun EntryCard(
    title: String,
    subtitle: String,
    meta: String,
    icon: ImageVector,
    onDelete: () -> Unit,
    modifier: Modifier = Modifier,
    onClick: (() -> Unit)? = null,
    details: (@Composable () -> Unit)? = null,
    action: (@Composable () -> Unit)? = null,
) {
    var expanded by remember { mutableStateOf(false) }
    val cardModifier = modifier.fillMaxWidth().let {
        if (onClick != null) it.clickable(onClick = onClick) else it
    }
    Card(modifier = cardModifier) {
        Column {
            Row(
                Modifier.padding(start = 16.dp, top = 12.dp, bottom = 12.dp, end = 4.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Icon(
                    icon,
                    contentDescription = null,
                    tint = MaterialTheme.colorScheme.primary,
                    modifier = Modifier.size(28.dp).padding(end = 12.dp),
                )
                Column(Modifier.weight(1f)) {
                    Text(title, style = MaterialTheme.typography.titleMedium)
                    if (subtitle.isNotBlank()) {
                        Text(
                            subtitle,
                            style = MaterialTheme.typography.bodyMedium,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }
                }
                Text(
                    meta,
                    style = MaterialTheme.typography.labelMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                IconButton(onClick = onDelete) {
                    Icon(
                        Icons.Filled.Delete,
                        contentDescription = stringResource(R.string.cd_delete),
                        tint = MaterialTheme.colorScheme.outline,
                    )
                }
            }
            if (details != null || action != null) {
                Row(
                    Modifier.padding(start = 8.dp, bottom = 4.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    if (details != null) {
                        TextButton(onClick = { expanded = !expanded }) {
                            Text(stringResource(R.string.action_details))
                            Icon(
                                if (expanded) Icons.Filled.ExpandLess else Icons.Filled.ExpandMore,
                                contentDescription = null,
                            )
                        }
                    }
                    if (action != null) action()
                }
                if (details != null && expanded) {
                    Box(Modifier.padding(start = 16.dp, end = 16.dp, bottom = 12.dp)) {
                        details()
                    }
                }
            }
        }
    }
}

/** A day divider in a history list: "Today", "Yesterday", or the date. */
@Composable
fun DayHeader(dayStartMillis: Long) {
    val context = LocalContext.current
    Text(
        dayLabel(context, dayStartMillis, System.currentTimeMillis()),
        style = MaterialTheme.typography.labelLarge,
        fontWeight = FontWeight.SemiBold,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
        modifier = Modifier.padding(top = 4.dp),
    )
}

/** The "+" button that opens a tracking screen's log form. */
@Composable
fun AddEntryFab(contentDescription: String, onClick: () -> Unit) {
    FloatingActionButton(onClick = onClick) {
        Icon(Icons.Filled.Add, contentDescription = contentDescription)
    }
}

/**
 * A tracking screen's log form, shown as a modal bottom sheet over the history
 * list (opened from the "+" button). The [content] scrolls if it outgrows the
 * sheet and stays clear of the keyboard.
 */
@Composable
fun AddEntrySheet(
    title: String,
    onDismiss: () -> Unit,
    content: @Composable ColumnScope.() -> Unit,
) {
    ModalBottomSheet(onDismissRequest = onDismiss) {
        Column(
            Modifier
                .verticalScroll(rememberScrollState())
                .imePadding()
                .padding(start = 16.dp, end = 16.dp, bottom = 24.dp),
        ) {
            Text(title, style = MaterialTheme.typography.titleMedium)
            content()
        }
    }
}

@Composable
fun EmptyHint(text: String, modifier: Modifier = Modifier) {
    Box(modifier = modifier.fillMaxWidth().padding(32.dp), contentAlignment = Alignment.Center) {
        Text(
            text,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }
}

@Composable
fun FieldLabel(text: String) {
    Text(
        text,
        style = MaterialTheme.typography.labelLarge,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
        modifier = Modifier.padding(top = 8.dp, bottom = 4.dp),
    )
}

@Composable
fun <T> ChoiceChips(
    options: List<T>,
    selected: T?,
    onSelect: (T) -> Unit,
    labelOf: (T) -> String,
    modifier: Modifier = Modifier,
) {
    FlowRow(
        modifier = modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        options.forEach { option ->
            FilterChip(
                selected = option == selected,
                onClick = { onSelect(option) },
                label = { Text(labelOf(option)) },
            )
        }
    }
}

@Composable
fun NumberField(
    label: String,
    value: String,
    onChange: (String) -> Unit,
    suffix: String? = null,
    modifier: Modifier = Modifier,
) {
    OutlinedTextField(
        value = value,
        onValueChange = { new -> onChange(new.filter { it.isDigit() }) },
        label = { Text(label) },
        suffix = suffix?.let { { Text(it) } },
        singleLine = true,
        keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
        modifier = modifier.fillMaxWidth(),
    )
}

@Composable
fun NotesField(value: String, onChange: (String) -> Unit, modifier: Modifier = Modifier) {
    OutlinedTextField(
        value = value,
        onValueChange = onChange,
        label = { Text(stringResource(R.string.field_notes_optional)) },
        modifier = modifier.fillMaxWidth(),
    )
}

/**
 * The confirmation shown before a destructive tap. Entries go for good and the
 * delete button sits within a thumb's width of the rest of a card, so a mis-tap
 * while holding a baby shouldn't cost you the log.
 */
@Composable
fun ConfirmDeleteDialog(
    onConfirm: () -> Unit,
    onDismiss: () -> Unit,
    title: String = stringResource(R.string.entry_delete_title),
    body: String = stringResource(R.string.entry_delete_body),
) {
    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text(title) },
        text = { Text(body) },
        confirmButton = {
            TextButton(onClick = onConfirm) {
                Text(
                    stringResource(R.string.action_delete),
                    color = MaterialTheme.colorScheme.error,
                )
            }
        },
        dismissButton = {
            TextButton(onClick = onDismiss) { Text(stringResource(R.string.action_cancel)) }
        },
    )
}

private fun zone(): ZoneId = ZoneId.systemDefault()

/**
 * Blocks days after today, for dates that record something that has already
 * happened. [DatePicker] works in UTC-midnight millis, so today's *local* date
 * is pinned at UTC midnight to compare against.
 */
private object PastOrToday : SelectableDates {
    override fun isSelectableDate(utcTimeMillis: Long): Boolean =
        utcTimeMillis <= LocalDate.now(zone()).atStartOfDay(ZoneOffset.UTC).toInstant().toEpochMilli()

    override fun isSelectableYear(year: Int): Boolean = year <= LocalDate.now(zone()).year
}

/**
 * A date field rendered as "label: date", opening the Material date picker.
 * Set [allowFuture] to false for a date that can only be in the past.
 */
@Composable
fun DatePickerField(
    label: String,
    millis: Long,
    onChange: (Long) -> Unit,
    allowFuture: Boolean = true,
) {
    var open by remember { mutableStateOf(false) }
    val context = LocalContext.current
    OutlinedButton(onClick = { open = true }, modifier = Modifier.fillMaxWidth()) {
        Text("$label: ${formatDate(context, millis)}")
    }
    if (open) {
        val state = rememberDatePickerState(
            initialSelectedDateMillis = millis,
            selectableDates = if (allowFuture) DatePickerDefaults.AllDates else PastOrToday,
        )
        DatePickerDialog(
            onDismissRequest = { open = false },
            confirmButton = {
                TextButton(onClick = {
                    state.selectedDateMillis?.let { picked ->
                        // Preserve the time-of-day from the previous value.
                        val prev = Instant.ofEpochMilli(millis).atZone(zone())
                        val newDate = Instant.ofEpochMilli(picked).atZone(ZoneId.of("UTC")).toLocalDate()
                        val combined = newDate.atTime(prev.toLocalTime()).atZone(zone()).toInstant().toEpochMilli()
                        onChange(combined)
                    }
                    open = false
                }) { Text(stringResource(R.string.action_ok)) }
            },
            dismissButton = {
                TextButton(onClick = { open = false }) { Text(stringResource(R.string.action_cancel)) }
            },
        ) {
            DatePicker(state = state)
        }
    }
}

@Composable
fun TimePickerField(label: String, millis: Long, onChange: (Long) -> Unit) {
    var open by remember { mutableStateOf(false) }
    OutlinedButton(onClick = { open = true }, modifier = Modifier.fillMaxWidth()) {
        Text("$label: ${formatTime(millis)}")
    }
    if (open) {
        val current = Instant.ofEpochMilli(millis).atZone(zone())
        val state = rememberTimePickerState(
            initialHour = current.hour,
            initialMinute = current.minute,
            is24Hour = true,
        )
        Dialog(onDismissRequest = { open = false }) {
            Card {
                Column(Modifier.padding(16.dp)) {
                    Text(label, style = MaterialTheme.typography.titleMedium)
                    Box(Modifier.padding(vertical = 16.dp)) { TimePicker(state = state) }
                    Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.End) {
                        TextButton(onClick = { open = false }) {
                            Text(stringResource(R.string.action_cancel))
                        }
                        TextButton(onClick = {
                            val newTime = LocalTime.of(state.hour, state.minute)
                            val combined = current.toLocalDate().atTime(newTime)
                                .atZone(zone()).toInstant().toEpochMilli()
                            onChange(combined)
                            open = false
                        }) { Text(stringResource(R.string.action_ok)) }
                    }
                }
            }
        }
    }
}

File: ios/Sprout/Sources/DesignSystem/Components.swift
import SproutData
import SwiftUI

// The shared views, from `ui/common/Components.kt`.
//
// Everything a tracking screen is built out of lives here, so the screens stay
// about *what* they log rather than how a card is spaced. Where SwiftUI has a
// better idiom than Compose the idiom wins — a sheet is a `.sheet`, a picker is
// a `DatePicker` — but the shape of each piece and the reason it exists are the
// same on both phones.

// MARK: - Structure

/// A section heading inside a screen.
struct SectionLabel: View {
    let text: String

    init(_ text: String) { self.text = text }

    var body: some View {
        Text(text)
            .font(.headline)
            .foregroundStyle(SproutColor.onSurface)
            .frame(maxWidth: .infinity, alignment: .leading)
    }
}

/// The label above a form field.
struct FieldLabel: View {
    let text: String

    init(_ text: String) { self.text = text }

    var body: some View {
        Text(text)
            .font(.subheadline.weight(.medium))
            .foregroundStyle(SproutColor.onSurfaceVariant)
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.top, Spacing.tight)
    }
}

/// What a history list shows before anything has been logged.
///
/// A sentence, not an illustration: the first thing a parent sees on this screen
/// is at 3 a.m. and should tell them what to do, not decorate the wait.
struct EmptyHint: View {
    let text: String

    init(_ text: String) { self.text = text }

    var body: some View {
        Text(text)
            .font(.callout)
            .foregroundStyle(SproutColor.onSurfaceVariant)
            .multilineTextAlignment(.center)
            .frame(maxWidth: .infinity)
            .padding(Spacing.section)
    }
}

/// A day divider in a history list: "Today", "Yesterday", or the date.
struct DayHeader: View {
    let dayStartMillis: Int64
    let now: Int64

    var body: some View {
        Text(SproutFormat.dayLabel(dayStartMillis, now: now).text)
            .font(.subheadline.weight(.semibold))
            .foregroundStyle(SproutColor.onSurfaceVariant)
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.top, Spacing.hairline)
    }
}

/// One figure on the dashboard — today's feeds, the sleep total, the nappy count.
struct StatCard: View {
    let label: String
    let value: String
    let systemImage: String

    var body: some View {
        VStack(alignment: .leading, spacing: Spacing.tight) {
            Image(systemName: systemImage)
                .font(.title3)
                .foregroundStyle(SproutColor.primary)
            Text(value)
                .font(.title2.weight(.semibold))
                .foregroundStyle(SproutColor.onSurface)
                // A long duration must shrink rather than wrap: the row of cards
                // has to stay one row at every text size.
                .lineLimit(1)
                .minimumScaleFactor(0.6)
            Text(label)
                .font(.caption)
                .foregroundStyle(SproutColor.onSurfaceVariant)
                .lineLimit(2)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(Spacing.regular)
        .background(SproutColor.surface, in: RoundedRectangle(cornerRadius: Radius.card))
    }
}

/// One logged entry in a history list.
///
/// `details` adds an expandable breakdown under the row; `action` puts a one-tap
/// shortcut beside it, for the state change that would otherwise mean opening
/// the editor.
struct EntryCard<Details: View, Action: View>: View {
    let title: String
    var subtitle: String = ""
    let meta: String
    let systemImage: String
    var onTap: (() -> Void)?
    let onDelete: () -> Void
    @ViewBuilder var details: () -> Details
    @ViewBuilder var action: () -> Action

    @State private var expanded = false

    private var hasDetails: Bool { Details.self != EmptyView.self }
    private var hasAction: Bool { Action.self != EmptyView.self }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            row
            if hasDetails || hasAction {
                HStack(spacing: Spacing.tight) {
                    if hasDetails {
                        Button {
                            withAnimation(.snappy) { expanded.toggle() }
                        } label: {
                            HStack(spacing: Spacing.hairline) {
                                Text(Str.t("action_details"))
                                Image(systemName: expanded ? "chevron.up" : "chevron.down")
                            }
                            .font(.footnote)
                        }
                        .buttonStyle(.plain)
                        .foregroundStyle(SproutColor.primary)
                    }
                    action()
                }
                .padding(.horizontal, Spacing.regular)
                .padding(.bottom, Spacing.snug)

                if hasDetails, expanded {
                    details()
                        .padding(.horizontal, Spacing.regular)
                        .padding(.bottom, Spacing.snug)
                }
            }
        }
        .background(SproutColor.surface, in: RoundedRectangle(cornerRadius: Radius.card))
        .contentShape(Rectangle())
        .onTapGesture { onTap?() }
    }

    private var row: some View {
        HStack(spacing: Spacing.snug) {
            Image(systemName: systemImage)
                .font(.title3)
                .foregroundStyle(SproutColor.primary)
                .frame(width: 28)

            VStack(alignment: .leading, spacing: 2) {
                Text(title)
                    .font(.body.weight(.medium))
                    .foregroundStyle(SproutColor.onSurface)
                if !subtitle.isEmpty {
                    Text(subtitle)
                        .font(.callout)
                        .foregroundStyle(SproutColor.onSurfaceVariant)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)

            Text(meta)
                .font(.caption)
                .foregroundStyle(SproutColor.onSurfaceVariant)

            Button(action: onDelete) {
                Image(systemName: "trash")
                    .foregroundStyle(SproutColor.outline)
            }
            .buttonStyle(.plain)
            .accessibilityLabel(Str.t("cd_delete"))
        }
        .padding(.leading, Spacing.regular)
        .padding(.trailing, Spacing.snug)
        .padding(.vertical, Spacing.snug)
    }
}

extension EntryCard where Details == EmptyView, Action == EmptyView {
    init(
        title: String,
        subtitle: String = "",
        meta: String,
        systemImage: String,
        onTap: (() -> Void)? = nil,
        onDelete: @escaping () -> Void
    ) {
        self.init(
            title: title,
            subtitle: subtitle,
            meta: meta,
            systemImage: systemImage,
            onTap: onTap,
            onDelete: onDelete,
            details: { EmptyView() },
            action: { EmptyView() }
        )
    }
}

// MARK: - Input

/// A row of single-choice chips.
///
/// Chips rather than a `Picker`: the options are the point of the form — which
/// breast, which nappy, where the baby slept — and a wheel or a menu hides them
/// behind a tap that a one-handed parent has to aim at.
struct ChoiceChips<Option: Hashable>: View {
    let options: [Option]
    @Binding var selection: Option?
    let label: (Option) -> String
    /// Whether tapping the selected chip clears it. On for the optional fields,
    /// where "not recorded" is a real answer and not a gap (BDR-0014).
    var allowsDeselection = false

    var body: some View {
        FlowLayout(spacing: Spacing.tight) {
            ForEach(options, id: \.self) { option in
                let isSelected = option == selection
                Button {
                    selection = (isSelected && allowsDeselection) ? nil : option
                } label: {
                    Text(label(option))
                        .font(.callout)
                        .padding(.horizontal, Spacing.snug)
                        .padding(.vertical, Spacing.tight)
                        .background(
                            isSelected ? SproutColor.primaryContainer : SproutColor.surface,
                            in: Capsule()
                        )
                        .overlay(
                            Capsule().stroke(
                                isSelected ? SproutColor.primary : SproutColor.outline.opacity(0.4),
                                lineWidth: 1
                            )
                        )
                        .foregroundStyle(
                            isSelected ? SproutColor.onPrimaryContainer : SproutColor.onSurface
                        )
                }
                .buttonStyle(.plain)
                .accessibilityAddTraits(isSelected ? [.isSelected] : [])
            }
        }
    }
}

/// Digits only, with an optional unit after the field.
struct NumberField: View {
    let label: String
    @Binding var text: String
    var suffix: String?

    var body: some View {
        HStack {
            TextField(label, text: $text)
                .keyboardType(.numberPad)
                .onChange(of: text) { _, new in
                    // Filtered rather than validated on submit: a stray letter
                    // from a predictive keyboard should never reach the model.
                    let digits = new.filter(\.isNumber)
                    if digits != new { text = digits }
                }
            if let suffix {
                Text(suffix).foregroundStyle(SproutColor.onSurfaceVariant)
            }
        }
        .padding(Spacing.snug)
        .background(SproutColor.surface, in: RoundedRectangle(cornerRadius: Radius.control))
    }
}

struct NotesField: View {
    @Binding var text: String

    var body: some View {
        TextField(Str.t("field_notes_optional"), text: $text, axis: .vertical)
            .lineLimit(1...4)
            .padding(Spacing.snug)
            .background(SproutColor.surface, in: RoundedRectangle(cornerRadius: Radius.control))
    }
}

/// Date and time on one row, which is how an entry is nearly always corrected:
/// the parent is logging a feed that happened twenty minutes ago.
struct DateTimeField: View {
    let label: String
    @Binding var millis: Int64

    private var date: Binding<Date> {
        Binding(
            get: { SproutFormat.date(from: millis) },
            set: { millis = SproutFormat.millis(from: $0) }
        )
    }

    var body: some View {
        DatePicker(label, selection: date, displayedComponents: [.date, .hourAndMinute])
            .font(.callout)
    }
}

// MARK: - Actions

/// The "+" that opens a tracking screen's log form.
struct AddEntryButton: View {
    let accessibilityLabel: String
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            Image(systemName: "plus")
                .font(.title2.weight(.semibold))
                .foregroundStyle(SproutColor.onPrimary)
                .frame(width: 56, height: 56)
                .background(SproutColor.primary, in: Circle())
                .shadow(color: .black.opacity(0.15), radius: 8, y: 4)
        }
        .buttonStyle(.plain)
        .accessibilityLabel(accessibilityLabel)
    }
}

/// The confirmation before a destructive tap.
///
/// Entries go for good, and the delete button sits within a thumb's width of the
/// rest of a card — a mis-tap while holding a baby should not cost the log.
extension View {
    func confirmDelete(
        isPresented: Binding<Bool>,
        title: String = Str.t("entry_delete_title"),
        message: String = Str.t("entry_delete_body"),
        onConfirm: @escaping () -> Void
    ) -> some View {
        confirmationDialog(title, isPresented: isPresented, titleVisibility: .visible) {
            Button(Str.t("action_delete"), role: .destructive, action: onConfirm)
            Button(Str.t("action_cancel"), role: .cancel) {}
        } message: {
            Text(message)
        }
    }
}

// MARK: - Layout

/// Chips wrap onto as many rows as they need.
///
/// SwiftUI has no `FlowRow`, and seven languages means a row of chips that fits
/// in English does not fit in German. Twenty lines of `Layout` rather than a
/// horizontal scroll view, because a chip a parent cannot see is a chip they
/// will not tap.
struct FlowLayout: Layout {
    var spacing: CGFloat = 8

    func sizeThatFits(proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) -> CGSize {
        let width = proposal.width ?? .infinity
        let rows = arrange(subviews: subviews, width: width)
        let height = rows.reduce(CGFloat.zero) { $0 + $1.height + spacing }
        return CGSize(width: width == .infinity ? rows.map(\.width).max() ?? 0 : width,
                      height: max(0, height - spacing))
    }

    func placeSubviews(
        in bounds: CGRect, proposal: ProposedViewSize, subviews: Subviews, cache: inout ()
    ) {
        var y = bounds.minY
        for row in arrange(subviews: subviews, width: bounds.width) {
            var x = bounds.minX
            for index in row.indices {
                let size = subviews[index].sizeThatFits(.unspecified)
                subviews[index].place(
                    at: CGPoint(x: x, y: y + (row.height - size.height) / 2),
                    proposal: ProposedViewSize(size)
                )
                x += size.width + spacing
            }
            y += row.height + spacing
        }
    }

    private struct Row {
        var indices: [Int] = []
        var width: CGFloat = 0
        var height: CGFloat = 0
    }

    private func arrange(subviews: Subviews, width: CGFloat) -> [Row] {
        var rows: [Row] = []
        var current = Row()

        for index in subviews.indices {
            let size = subviews[index].sizeThatFits(.unspecified)
            let needed = current.indices.isEmpty ? size.width : current.width + spacing + size.width
            if needed > width, !current.indices.isEmpty {
                rows.append(current)
                current = Row()
            }
            current.width = current.indices.isEmpty ? size.width : current.width + spacing + size.width
            current.height = max(current.height, size.height)
            current.indices.append(index)
        }
        if !current.indices.isEmpty { rows.append(current) }
        return rows
    }
}
